table/tables/tables.go

```diff
@@ -187,115 +211,100 @@ func (t *Table) FirstKey() kv.Key {
+			if err != nil {
+				return errors.Trace(err)
+			}
+			cmp, err := oldData[col.Offset].CompareDatum(ctx.GetSessionVars().StmtCtx, value)
+			if err != nil {
+				return errors.Trace(err)
 			}
-			currentData[i] = defaultVal
+			if cmp != 0 {
+				value = oldData[col.Offset]
+			}
+		} else {
+			value = newData[col.Offset]
+		}
+		if !t.canSkip(col, value) {
+			colIDs = append(colIDs, col.ID)
+			row = append(row, value)
+		}
+		if shouldWriteBinlog(ctx) && !t.canSkipUpdateBinlog(col, value) {
+			binlogColIDs = append(binlogColIDs, col.ID)
+			binlogOldRow = append(binlogOldRow, oldData[col.Offset])
+			binlogNewRow = append(binlogNewRow, value)
 		}
-		colIDs = append(colIDs, col.ID)
 	}
-	// Set new row data into KV.
+
 	key := t.RecordKey(h)
-	value, err := tablecodec.EncodeRow(currentData, colIDs, ctx.GetSessionVars().GetTimeZone())
+	value, err := tablecodec.EncodeRow(row, colIDs, ctx.GetSessionVars().GetTimeZone())
 	if err != nil {
 		return errors.Trace(err)
 	}
 	if err = bs.Set(key, value); err != nil {
 		return errors.Trace(err)
 	}
-
-	// rebuild index
-	if err = t.rebuildIndices(bs, h, touched, oldData, currentData); err != nil {
-		return errors.Trace(err)
-	}
-
-	err = bs.SaveTo(txn)
-	if err != nil {
+	if err = bs.SaveTo(txn); err != nil {
 		return errors.Trace(err)
 	}
 	if shouldWriteBinlog(ctx) {
-		t.addUpdateBinlog(ctx, h, oldData, value, colIDs)
+		t.addUpdateBinlog(ctx, binlogOldRow, binlogNewRow, binlogColIDs)
 	}
 	return nil
 }
 
-func (t *Table) setOnUpdateData(ctx context.Context, touched map[int]bool, data []types.Datum) error {
-	ucols := table.FindOnUpdateCols(t.WritableCols())
-	for _, col := range ucols {
-		if !touched[col.Offset] {
-			value, err := expression.GetTimeValue(ctx, expression.CurrentTimestamp, col.Tp, col.Decimal)
-			if err != nil {
-				return errors.Trace(err)
+func (t *Table) rebuildIndices(rm kv.RetrieverMutator, h int64, touched []bool, oldData []types.Datum, newData []types.Datum) error {
+	for _, idx := range t.DeletableIndices() {
+		for _, ic := range idx.Meta().Columns {
+			if touched[ic.Offset] {
+				oldVs, err := idx.FetchValues(oldData)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if t.removeRowIndex(rm, h, oldVs, idx); err != nil {
+					return errors.Trace(err)
+				}
+				break
 			}
-			data[col.Offset] = value
-			touched[col.Offset] = true
 		}
 	}
-	return nil
-}
-
-// composeNewData fills untouched columns with original values.
-// TODO: consider col state
-func (t *Table) composeNewData(touched map[int]bool, newData []types.Datum, oldData []types.Datum) {
-	for i, od := range oldData {
-		if touched[i] {
-			continue
-		}
-		newData[i] = od
-	}
-	return
-}
-
-func (t *Table) rebuildIndices(rm kv.RetrieverMutator, h int64, touched map[int]bool, oldData []types.Datum, newData []types.Datum) error {
-	for _, idx := range t.Indices() {
-		idxTouched := false
+	for _, idx := range t.WritableIndices() {
```

Comment:
Previously check the index is writable or not is done in `buildIndexForRow`. I moved that here for more easily to understand.